### PR TITLE
Add the chainscript code

### PR DESCRIPTION
**Pull request by iljakuklic**
This moves the code from the `chainscript` repo in the main node repo. Integration of the scripting capabilities into the node will be a separate PR. For now, the extent of integration reduces to the ability to run chainscript tests together with all the other tests.

The code in `libs/chainscript` folder has already been reviewed, you only need to pay attention to changes outside of that folder, which are all in the last commit.

### CODE REVIEW

**Review by altonen — APPROVED** [review]

**Review by muursh — APPROVED** [review]
Looks good

File: libs/chainscript/src/context.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+    /// Extract a pubkey and check it is in the correct format.
+    fn parse_pubkey(&self, pk: &[u8]) -> Option<Self::Public>;
+
+    /// Verify signature.
+    fn verify_signature(&self, sig: &Self::Signature, pk: &Self::Public, subscript: &[u8]) -> bool;
+
+    /// Check absolute time lock.
+    fn check_lock_time(&self, _lock_time: i64) -> bool {
+        false
```
[the remark this answers is not in the document]
Comment (muursh) [reply libs/chainscript/src/context.rs:61]: Okay this has come up on slack today and we've decided to stick with 1 min blocks

File: libs/chainscript/src/context.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+/// * This interface is currently rather monolithic. It may be sensible to break it down into
+///   multiple smaller ones for greater flexibility in the future.
+/// * The interface is not fully finalized. It is likely to change as new requirements come in. E.g.
+///   it may be currently too limited to support signature batching.
+pub trait Context {
+    /// Maximum number of bytes pushable to the stack
+    const MAX_SCRIPT_ELEMENT_SIZE: usize = 520;
+
+    /// Maximum number of public keys per multisig
+    const MAX_PUBKEYS_PER_MULTISIG: usize;
```
[the remark this answers is not in the document]
Comment (iljakuklic) [reply libs/chainscript/src/context.rs:39]: This particular setting is not implemented at this point, it's on my todo list

File: libs/chainscript/src/context.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    }
+}
+
+// A test context implementation.
+// Used for testing and as an example of what a Context might look like.
+#[cfg(any(test, feature = "testcontext"))]
+pub mod testcontext {
+
+    use super::*;
+    use crate::util::sha256;
```
[the remark this answers is not in the document]
Comment (iljakuklic) [reply libs/chainscript/src/context.rs:86]: `util` just reexports the `sha2_256` function from the `sp_io` package:

```rust
pub use sp_io::hashing::sha2_256 as sha256;
```

I guess it's ultimately taken from the crate you linked.

File: libs/chainscript/src/context.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (muursh) [reply libs/chainscript/src/context.rs:39]: Yeah don't worry it's a while away in the node. I technically have the code to handle it but it's not integrated with substrate at all right now

**Pull request merged**